test/integration/test_celery_user_rate_limit.py

```diff
@@ -0,0 +1,169 @@
+from galaxy.model.database_utils import sqlalchemy_engine
+from galaxy.util import ExecutionTimer
+from galaxy_test.driver.driver_util import init_database
+from galaxy_test.driver.integration_util import IntegrationTestCase
+
+
+@galaxy_task(bind=True)
+def mock_user_id_task(self, task_user_id: int):
+    return task_user_id
+
+
+@lru_cache()
+def sqlite_url():
+    path = tempfile.NamedTemporaryFile().name
+    dburl = f"sqlite:///{path}"
+    init_database(dburl)
+    return dburl
+
+
+@lru_cache()
+def setup_users(dburl: str, num_users: int = 2):
+    """
+    Setup test users in galaxy_user table with user id's starting from 2 because
+    we assume there always exists a user with id = 1.
+    This is because the new celery_user_rate_limit table has
+    a user_id with a foreign key pointing to galaxy_user table.
+    """
+    expected_user_ids = [i for i in range(2, num_users + 1)]
+    with sqlalchemy_engine(dburl) as engine:
+        with engine.begin() as conn:
+            found_user_ids = conn.scalars(
+                text("select id from galaxy_user where id between 1 and :high"), {"high": num_users}
+            ).all()
+            if len(expected_user_ids) > len(found_user_ids):
+                user_ids_to_add = set(expected_user_ids).difference(found_user_ids)
+                for user_id in user_ids_to_add:
+                    conn.execute(
+                        text("insert into galaxy_user(id, active, email, password) values (:id, :active, :email, :pw)"),
+                        [{"id": user_id, "active": True, "email": "e", "pw": "p"}],
+                    )
+
+
+class TestCeleryUserRateLimitIntegration(IntegrationTestCase):
+    """
+    Test feature that limits the number of celery task executions
+    per user per second. This is implemented by the celery_user_rate_limit
+    config parameter whose default value is 0.0 meaning no rate limit.
+    For each of a set of users it submits runs a task num_calls times.
+    For the test to succeed the total duration of completion of all task
+    executions should be equal to the tasks_per_user_per_sec times
+    num_calls-1. This is because tasks for different users should be
+    executed in parallel. In addition, we verify that the total
+    duration for execution of tasks for each user equals the total
+    task duration, again because tasks for different users should
+    run in parallel.
+    """
+
+    def setUp(self):
+        super().setUp()
+
+    def _test_mock_pass_user_id_task(self, users: Iterable[int], num_calls: int, tasks_per_user_per_sec: float):
+        expected_duration: float
+        if tasks_per_user_per_sec == 0.0:
+            expected_duration = 0.0
+            expected_duration_lbound = 0.0
+        else:
+            secs_between_tasks_per_user = 1 / tasks_per_user_per_sec
+            expected_duration = secs_between_tasks_per_user * (num_calls - 1)
+            expected_duration_lbound = expected_duration - 4
```

Comment:
I am a little confused about the calculations here. Can you explain, please? (it might be obvious - sorry if that's the case!)
- What is `4` and why is it hardcoded here? (I see it adjusts the lower/higher bounds)
- `secs_between_tasks_per_user = 1 / tasks_per_user_per_sec`: why is that seconds between tasks?

Reply:
To illustrate, assume tasks_per_user_per_sec = 2. This means that even though the user may submit thousands of tasks to execute at the same time, the user is allowed to execute at most 2 task per second or at most 120 tasks per minute. In this case secs_between_tasks_per_user = 1/2 = .5 sec.
In order to achieve this we need to schedule the tasks via retry to execute .5 seconds apart.

In order to verify that the rate limiting is working as expected I verify that the total time it takes to execute all of the tasks submitted at once in the test execute within a certain period of time. So to continue the above example, if the user submits 120 tasks at once, I need to verify that it takes 60 seconds to complete execution of all the tasks, i.e. secs_between_tasks_per_user(.5) * num_calls(120) = 60 seconds.
However, because I cannot be guaranteed that it will be exactly 60 seconds I make it a range, i.e. 60 + or - 4 seconds. Of course, there is nothing magic about choosing 4 as the degree of error. I came up with it via trial and error. I suppose I could assign it to a class variable.

Reply:
Thank you - this is very helpful! I think I get it now :) 
The value `4` is fine as is.